planet/src/main/java/com/google/planet/data/ItineraryGenerator.java

```diff
@@ -47,74 +51,64 @@ public Itinerary generateItinerary(List<Event> events) {
+                    TimeRange.fromStartDuration(start, event.getDurationInMinutes()));
+                items.add(item);
 
-            for (int i = 0; i < distanceMatrix.rows.length; i++) {
-                DistanceMatrixRow currentRow = distanceMatrix.rows[i];
-                for (int j = 0; j < currentRow.elements.length; j++) {
-                    DistanceMatrixElement currentElement = currentRow.elements[j];
-                    if (currentElement.duration != null) {
-                        travelTimeGraph[i][j] = currentElement.duration;
-                    } else {
-                        errorMessage = "Oops, one of your addresses is invalid, please use a valid address.";
-                    }
-                }
+                // Update the next event's start time
+                start += event.getDurationInMinutes() + getTravelDurationInMinutes(directionsLegs[i].duration);
+            }else {
+                errorMessage = "Sorry, you have too many events in a day!";
+                return Arrays.asList();
             }
-            return travelTimeGraph;
-        } catch (Exception e) {
-            errorMessage = "Oops, one of your addresses is invalid, please use a valid address.";
-            return travelTimeGraph;
         }
+        return items;
     }
 
-    private String[] getListOfAddresses(List<Event> events) {
-        String[] addressList = events.stream()
-                                .map(event -> event.getAddress())
-                                .toArray(String[]::new);
-        return addressList;
-    }
-
-    private int getTravelDurationInMinutes(Duration duration) {
-        int travelDurationInMinutes = (int) Math.round(duration.inSeconds / 60);
-        return travelDurationInMinutes;
-    }
-
-    // Function that creates an itinerary by scheduling each event in order.
-    private List<ItineraryItem> scheduleItineraryInOrder(List<Event> events, int openningTime, int endingTime, 
-                                                        Duration[][] travelTimeGraph) {
+    // Function that creates an optimized itinerary
+    private List<ItineraryItem> scheduleOptimizedItinerary(List<Event> events, int openningTime, int endingTime) {
         List<ItineraryItem> items = new ArrayList<>();
+        boolean optimized = true;
+        DirectionsRoute directionsRoute = getDirectionsRoute(events, optimized);
+        if (errorMessage != null) { 
+            return items;
+        }
+        DirectionsLeg[] directionsLegs = directionsRoute.legs;
         int start = openningTime; 
+        int eventIndex = 0;
         for (int i = 0; i < events.size(); i++){
-            Event event = events.get(i);
-
+            // waypointOrder gives the ordering of the wayPoints, excluding the origin.
+            // Therefore, waypointOrder of [2, 0, 1] will correspond to event 3, 1, and 2, etc.
+            if (i != 0) {
+                eventIndex = directionsRoute.waypointOrder[i-1] + 1;
```

Comment:
nit: you might want to pull out the zero index case before the for loop since it's a special case.

Something like:
- add zero index event and increment start time
- for i = 1..events.size()
  - loop

I'll leave it up to you to decide which seems most readable to someone unfamiliar with the code.

Reply:
Okay!